### PR TITLE
Fix --insecure-registry not working in some circumstances for #775

### PR DESCRIPTION
**Pull request by Zheaoli**
Signed off by: Zheao Li <me@manjusaka.me>

#775

### CODE REVIEW

File: cmd/nerdctl/login.go [hunk 1]
```diff
@@ -287,7 +289,14 @@ func tryLoginWithRegHost(ctx context.Context, rh docker.RegistryHost) error {
 		}
 		res, err := ctxhttp.Do(ctx, rh.Client, req)
 		if err != nil {
-			return fmt.Errorf("failed to call rh.Client.Do: %w", err)
+			// In some circumstance (e.g. people just use 80 port to support pure http), the error will contain message like "dial tcp <port>: connection refused"
+			if (imgutil.IsErrHTTPResponseToHTTPSClient(err) || imgutil.IsErrConnectionRefused(err)) && insecure {
```
Comment (Junnplus) [inline cmd/nerdctl/login.go:293]: If the registry service does not support HTTPS and the user does not specify `--insecure-registry`, we should alert them, like https://github.com/containerd/nerdctl/blob/v0.16.1/cmd/nerdctl/push.go#L197-L201

File: cmd/nerdctl/login.go [hunk 2]
```diff
@@ -256,7 +256,7 @@ func loginClientSide(ctx context.Context, cmd *cobra.Command, auth types.AuthCon
 	return "", err
 }
 
-func tryLoginWithRegHost(ctx context.Context, rh docker.RegistryHost) error {
+func tryLoginWithRegHost(ctx context.Context, rh docker.RegistryHost, insecure bool) error {
```
Comment (Junnplus) [inline cmd/nerdctl/login.go:259]: `insecure` doesn't seem to be used?

**Review by Junnplus — APPROVED** [review]
I don't know why the logic related to login was removed, other LGTM.

**Comment by Zheaoli** [comment]
> I don't know why the logic related to login was removed, other LGTM.

Because I find the TestPushInsecureWithLogin will fail sometimes, not every time if I add plain HTTP logic for login action. Such as https://github.com/containerd/nerdctl/runs/5122259173?check_suite_focus=true. So I choose to just support plain HTTP for pull and push action, and I will make another PR for login action if I have figure out what's happen

File: pkg/imgutil/imgutil.go [hunk 3]
```diff
@@ -200,6 +201,13 @@ func IsErrHTTPResponseToHTTPSClient(err error) bool {
 	return strings.Contains(err.Error(), unexposed)
 }
 
+// IsErrConnectionRefused return whether err is
+// "connect: connection refused"
+func IsErrConnectionRefused(err error) bool {
+	const errMessage = "connect: connection refused"
+	return strings.Contains(err.Error(), errMessage)
```
Comment (AkihiroSuda) [inline pkg/imgutil/imgutil.go:208]: Can't we use `errors.Is`?

**Review by AkihiroSuda — APPROVED** [review]
Thanks

File: pkg/imgutil/imgutil.go (hunk 3 above)
Comment (Zheaoli) [reply pkg/imgutil/imgutil.go:208]: I think I can make a single PR to make all error message checks like `errors.Is`

**Pull request merged**